src/mlpack/methods/lsh/lshmodel_impl.hpp

```diff
@@ -0,0 +1,610 @@
+using std::cout;
+using std::flush;
+using std::endl;
+
+namespace mlpack {
+namespace neighbor {
+
+// Constructor sets variables and trains the object.
+template <typename SortPolicy, typename ObjectiveFunction>
+LSHModel<SortPolicy, ObjectiveFunction>::
+LSHModel(const arma::mat &referenceSet,
+         const double sampleSize,
+         const size_t k)
+{
+  // We don't own the set - we just point to it.
+  ownsSet = false;
+  this->referenceSet = &referenceSet;
+
+  Train(referenceSet, sampleSize, k);
+}
+
+// Destructor must de-allocate any referenceSet and LSHSearch objects we own.
+template <typename SortPolicy, typename ObjectiveFunction>
+LSHModel<SortPolicy, ObjectiveFunction>::~LSHModel()
+{
+  if (ownsSet)
+    delete referenceSet;
+};
+
+// Trains the object.
+template <typename SortPolicy, typename ObjectiveFunction>
+void LSHModel<SortPolicy, ObjectiveFunction>::Train(
+    const arma::mat &referenceSet,
+    const double sampleRate,
+    const size_t k)
+{
+  // Sanity check - sample rate must be in (0, 1].
+  if (sampleRate > 1 || sampleRate <= 0)
+    Log::Fatal << "Sampling rate must be floating point number in (0, 1]"
+        << std::endl;
+
+  // Update the object's max K value information.
+  maxKValue = k;
+
+  // Save pointer to training set.
+  this->referenceSet = &referenceSet;
+
+  // Step 1. Select a random sample of the dataset. We will work with only that
+  // sample.
+  arma::vec sampleHelper(referenceSet.n_cols, arma::fill::randu);
+
+  // Keep a sample of the dataset: We have uniformly random numbers in [0, 1],
+  // so we expect about N*sampleRate of them to be in [0, sampleRate).
+  arma::mat sampleSet = referenceSet.cols(
+        arma::find(sampleHelper < sampleRate));
+  // Shuffle to be impartial (in case dataset is sorted in some way).
+  sampleSet = arma::shuffle(sampleSet);
+  const size_t numSamples = sampleSet.n_cols; // Points in sampled set.
+
+  Log::Info << "Training model with " << numSamples << " points in sample set."
+    << std::endl;
+
+  // Step 2. Compute all-vs-all distances of points in the sample.
+  // The distance matrix is symmetric, so we only compute elements above the
+  // diagonal. There are (N * (N - 1)) / 2 such elements.
+  Timer::Start("pairwise_distances");
+  arma::vec distances(numSamples * (numSamples - 1) / 2);
+  size_t d = 0; // Index of where to store next.
+  for (size_t i = 0; i < numSamples; ++i)
+    for (size_t j = i + 1; j < numSamples; ++j)
+      distances(d++) = metric::EuclideanDistance::Evaluate(
+          sampleSet.unsafe_col(i), sampleSet.unsafe_col(j));
+  Log::Info << "Computed " << d << " pointwise distances." << std::endl;
+  Timer::Stop("pairwise_distances");
+
+  // Step 3. Estimate statistics of these distances: log(mean(d)), mean(log(d)),
+  // mean(d).
+  distances = arma::pow(distances, 2);
+  this->meanDist = arma::mean(distances);
+  this->logMeanDist = std::log(meanDist);
+  this->meanLogDist = arma::mean(arma::log(distances));
```

Comment:
Even if you replace the geometric mean calculation below, this still causes problems in `Predict()`, because any zero values passed to the Gamma distribution training will cause it to fail, which makes sense because the PDF of a Gamma distribution (with any parameters) is always 0 at x = 0.

So basically what we have come upon here is a flaw in the paper: this entire strategy of modeling the squared distances as a gamma distribution is only applicable if there are no duplicate points in the dataset.  If we wanted this LSH tuning model to be able to handle duplicate points, we would need to pick a new distribution that handled zero distances and rederive everything in the paper.  But... that's not really what I want to do (if you want to feel free)...

So instead it seems like our best option is just to find a hack.  Here is the idea I went with to make my little test program work: take the smallest nonzero distance from `distances`, and use some small factor of that to replace all the 0s:

```
// Find the smallest nonzero distance.
double smallest = DBL_MAX;
for (size_t i = 0; i < distances.n_elem; ++i)
  if (distances[i] > 0.0 && distances[i] < DBL_MAX)
    smallest = distances[i];

// Replace any zero distances, since they break the gamma distribution.
for (size_t i = 0; i < distances.n_elem; ++i)
  if (distances[i] == 0.0)
    distances[i] = 1e-5 * smallest; // 1e-5 selected arbitrarily.
```

Now the distances are all nonzero, though we have probably affected the trained distribution somewhat.  It might be useful to output a warning if there are any duplicate points in the distances, noting that this affects the correctness of the predictions.  (I'm not sure how it affects the correctness.)